local_vars/local_vars.cc

```diff
@@ -226,35 +232,246 @@ class LocalNameInserter {
+        ast::Array::ENTRY_store posArgsEntries;
+        ast::Hash::ENTRY_store kwArgKeyEntries;
+        ast::Hash::ENTRY_store kwArgValueEntries;
+
+        // ...if we hit a splat along the way, however, we'll build it into an expression that evaluates to an array
+        // (or a hash, where keyword args are concerned). Otherwise, posArgsArray (resp. kwArgsHash) will be null.
+        ast::ExpressionPtr posArgsArray;
+        ast::ExpressionPtr kwArgsHash;
+
+        // We'll also look for the block arg, which should always be present at the end of the args.
+        ast::ExpressionPtr blockArg;
+
+        for (auto arg : scopeStack.back().args) {
+            ENFORCE(!blockArg, "Block arg was not in final position");
+
+            if (arg.flags.isPositional()) {
+                ENFORCE(kwArgKeyEntries.empty(), "Saw positional arg after keyword arg");
+                ENFORCE(!kwArgsHash, "Saw positional arg after keyword splat");
+
+                posArgsEntries.emplace_back(ast::make_expression<ast::Local>(original.loc, arg.arg));
+            } else if (arg.flags.isPositionalSplat()) {
+                ENFORCE(!posArgsArray, "Saw multiple positional splats");
+                ENFORCE(kwArgKeyEntries.empty(), "Saw positional splat after keyword arg");
+                ENFORCE(!kwArgsHash, "Saw positional splat after keyword splat");
+
+                posArgsArray = ast::MK::Splat(original.loc, ast::make_expression<ast::Local>(original.loc, arg.arg));
+                if (!posArgsEntries.empty()) {
+                    posArgsArray = ast::MK::Send1(original.loc, ast::MK::Array(original.loc, std::move(posArgsEntries)),
+                                                  core::Names::concat(), std::move(posArgsArray));
+                    posArgsEntries.clear();
                 }
+            } else if (arg.flags.isKeyword()) {
+                ENFORCE(!kwArgsHash, "Saw keyword arg after keyword splat");
+
+                kwArgKeyEntries.emplace_back(ast::MK::Literal(
+                    original.loc, core::make_type<core::LiteralType>(core::Symbols::Symbol(), arg.arg._name)));
+                kwArgValueEntries.emplace_back(ast::make_expression<ast::Local>(original.loc, arg.arg));
+            } else if (arg.flags.isKeywordSplat()) {
+                ENFORCE(!kwArgsHash, "Saw multiple keyword splats");
+
+                // TODO(aprocter): is it necessary to duplicate the hash here?
+                kwArgsHash =
+                    ast::MK::Send1(original.loc, ast::MK::Constant(original.loc, core::Symbols::Magic()),
+                                   core::Names::toHashDup(), ast::make_expression<ast::Local>(original.loc, arg.arg));
+                if (!kwArgKeyEntries.empty()) {
+                    // TODO(aprocter): it might make more sense to replace this with an InsSeq that calls
+                    // <Magic>::<merge-hash>, which is what's done in the desugarer.
```

Comment:
What's the plan for these two `TODO(aprocter)`'s? did you want to address those in this PR?

Reply:
I think they're okay for now, but they might not be the best choice perf-wise. Happy to address before merge, open a ticket, what-have-ya.